File: src/main/kotlin/uk/ac/ic/matterialize/matting/FaceDetectionMatter.kt
package uk.ac.ic.matterialize.matting

import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.MatOfRect
import org.opencv.core.Scalar
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc.COLOR_BGR2GRAY
import org.opencv.imgproc.Imgproc.cvtColor
import org.opencv.objdetect.CascadeClassifier

class FaceDetectionMatter() : Matter {
    private val classifier = CascadeClassifier("src/main/resources/haarcascade_frontalface_default.xml")

    override fun foregroundMask(videoFrame: Mat): Mat {
        val mask = Mat(videoFrame.size(), CvType.CV_8U, Scalar(0.0))
        val gray = Mat()
        cvtColor(videoFrame, gray, COLOR_BGR2GRAY)
        val faces = MatOfRect()
        classifier.detectMultiScale(
            gray,
            faces,
            1.05,
            8,
            0,
            Size(50.0, 50.0),
            Size(150.0, 150.0)
        )

        faces.toArray().forEach {
            ((it.x - it.width / 2) until it.x + (3 * it.width) / 2).forEach { i ->
                (it.y..mask.height()).forEach { j ->
                    mask.put(j, i, 255.0)
                }
            }
        }

        return mask
    }
}

File: c++/src/matting/face_detection_matter.cpp
#include "face_detection_matter.hpp"

#include <vector>

#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>

cv::Mat FaceDetectionMatter::background_mask(const cv::Mat &video_frame) {
  cv::Mat mask(video_frame.size(), CV_8U, cv::Scalar(0.0));
  cv::Mat gray{};

  cv::cvtColor(video_frame, gray, cv::COLOR_BGR2GRAY);

  std::vector<cv::Rect> faces{};

  classifier.detectMultiScale(
      gray, faces, 1.05, 8, 0, cv::Size(50.0, 50.0), cv::Size(150.0, 150.0));

  for (cv::Rect &rect : faces) {
    for (int i{rect.x - rect.width / 2}; i < rect.x + (3 * rect.width) / 2;
         ++i) {
      for (int j{rect.y}; j <= mask.size().height; ++j) {
        mask.at<cv::Vec2d>(j, i) = 255.0;
      }
    }
  }

  return mask;
}
